### PR TITLE
Incorporate multi selection menus!

### PR DESCRIPTION
**Pull request by wishdasher**

### CODE REVIEW

**Comment by jeresig** [comment]
Deploy preview for *wonder-blocks* ready!

Built with commit 50281475752d42281c11cd56343fc193d3f42963

https://deploy-preview-169--wonder-blocks.netlify.com

File: packages/wonder-blocks-dropdown/components/multi-select-menu.js [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+                    }
+                    selected={selectedItems.includes(item.value)}
+                    value={item.value}
+                    variant={"checkbox"}
+                />
+            );
+        });
+
+        if (displaySelectionShortcuts) {
+            // TODO(sophie): translate for i18n
```
Comment (kevinbarabash) [inline packages/wonder-blocks-dropdown/components/multi-select-menu.js]: I think as long as we have a version of `i18n._()` that just returns the string passed in we should be good.  @jeresig that should work right?  We should probably add a function to `wonder-blocks-core`.  That can be a separate PR.

File: packages/wonder-blocks-dropdown/components/multi-select-menu.js [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+            light,
+            placeholder,
+            selectItemType,
+            selectedItems,
+            style,
+        } = this.props;
+
+        const {open} = this.state;
+
+        // TODO(sophie): figure out how to configure plurals for i18n
```
Comment (kevinbarabash) [inline packages/wonder-blocks-dropdown/components/multi-select-menu.js]: We'll need a copy of the `i18n.ngettext()` function from `i18n.js` in `webapp`.  Probably best to add it to `wonder-blocks-core`... or maybe we can create an `i18n` package.  This can be a separate PR.

File: packages/wonder-blocks-dropdown/components/multi-select-menu.md [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+    }
+
+    handleChanges(changes) {
+        console.log("changes happened!");
+        const {selectedItems} = this.state;
+        changes.forEach((change) => {
+            if (change.selected) {
+                selectedItems.push(change.value);
+            } else {
+                selectedItems.splice(selectedItems.indexOf(change.value), 1);
```
Comment (kevinbarabash) [inline packages/wonder-blocks-dropdown/components/multi-select-menu.md]: If `change` was also an array then this would become:
```
this.setState({
   selectedItems: changes,
});
```

File: packages/wonder-blocks-dropdown/components/multi-select-menu.js [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+                    disabled={selectedItems.length === 0}
+                    indent={true}
+                    onClick={() => this.handleSelectNone()}
+                />
+            );
+
+            const separator = <SeparatorItem key={"selectOptSep"} />;
+
+            // Add options at beginning, add a separator
+            menuItems.unshift([selectAll, selectNone, separator]);
```
Comment (kevinbarabash) [inline packages/wonder-blocks-dropdown/components/multi-select-menu.js]: Maybe move the `displaySelectionShortcuts` block above the call to `const menuItems = items.map((item, index) => {`.  This avoids `unshift()` which can be bad for perf... in this situation not a huge deal.  Maybe something like:
```
const menuItems = [...this.maybeGetShortcuts(), ...this.getMenuitems()];
```

File: packages/wonder-blocks-dropdown/components/multi-select-menu.js [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+            />
+        );
+    }
+}
+
+const styles = StyleSheet.create({
+    // This is to add extra space on top of the menu options to separate the
+    // options from the opener component.
+    menuTopSpace: {
+        top: 48,
```
Comment (kevinbarabash) [inline packages/wonder-blocks-dropdown/components/multi-select-menu.js]: This is b/c the opener is `40` and the space between the opener and the menu is `8`.

File: packages/wonder-blocks-dropdown/components/multi-select-menu.md [hunk 6]
```diff
@@ -0,0 +1,241 @@
+The multi select menu allows the selection of multiple items. Clients are
+responsible for keeping track of the selected items in the menu.
+
+TODO(sophie): keyboard navigation and dismissal
+
+The first menu starts with nothing selected and has no selection shortcuts.
+There is also a disabled item, and this menu has a set width.
+
+The second menu starts with one item selected and has selection shortcuts for
+select all and select none. The second menu does not have a predefined placeholder.
```
Comment (kevinbarabash) [inline packages/wonder-blocks-dropdown/components/multi-select-menu.md]: I noticed that the items in these menus have white borders on the left/right whereas the ones in the SingleSelectMenu do not.  These shouldn't have any white borders on the left/right either.

File: packages/wonder-blocks-dropdown/components/multi-select-menu.md [hunk 7]
```diff
@@ -0,0 +1,241 @@
+The multi select menu allows the selection of multiple items. Clients are
+responsible for keeping track of the selected items in the menu.
+
+TODO(sophie): keyboard navigation and dismissal
+
+The first menu starts with nothing selected and has no selection shortcuts.
+There is also a disabled item, and this menu has a set width.
```
Comment (kevinbarabash) [inline packages/wonder-blocks-dropdown/components/multi-select-menu.md]: One of the examples should have a lot of choices so that we can show that the menu scrolls.

File: packages/wonder-blocks-dropdown/components/multi-select-menu.md [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+        />;
+    }
+}
+
+<View style={[styles.row]}>
+    <ExampleNoneSelected />
+    <View style={[styles.strutLike]} />
+    <ExampleWithShortcuts />
+    <View style={[styles.strutLike]} />
+    <ExampleNoneSelected disabled={true} />
```
Comment (kevinbarabash) [inline packages/wonder-blocks-dropdown/components/multi-select-menu.md]: Can you move these into separate examples?  It will keep the code snippet shorter and easier for devs to see what's going on.

File: packages/wonder-blocks-dropdown/components/single-select-menu.md [hunk 9]
```diff
@@ -197,15 +197,15 @@ class LightRightAlignedExample extends React.Component {
             items={[
                 {
                     label: "Regular milk tea with boba",
-                    value: "regular"
+                    value: "regular",
                 },
                 {
                     label: "Wintermelon milk tea with boba",
-                    value: "wintermelon"
+                    value: "wintermelon",
```
Comment (kevinbarabash) [inline packages/wonder-blocks-dropdown/components/single-select-menu.md]: This sounds really yummy.

File: packages/wonder-blocks-dropdown/utils/types.js [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+//     href?: string,
+//     /** Whether to use client-side navigation. */
+//     clientNav?: boolean,
+//     /** Callback on the action. */
+//     onClick?: () => void,
+// };
+
+export type SelectItemProps = {
+    type: "select",
+    /** Whether this item is disabled. Default false. */
```
Comment (kevinbarabash) [inline packages/wonder-blocks-dropdown/utils/types.js]: super nit: `/* ... */`

File: packages/wonder-blocks-dropdown/utils/types.js [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+    label: string,
+    /** Value of this item. Treat as a key. */
+    value: string,
+    /** Optional extra callback. Passes whether this item is selected. */
+    onClick?: (selected: boolean) => void,
+};
+
+// export type SeparatorProps = {
+//     type: "separator",
+// };
```
Comment (kevinbarabash) [inline packages/wonder-blocks-dropdown/utils/types.js]: Should this be used somewhere since we do have separators in some of the examples?

File: packages/wonder-blocks-dropdown/components/multi-select-menu.js [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+            style,
+        } = this.props;
+
+        const {open} = this.state;
+
+        // TODO(sophie): figure out how to configure plurals for i18n
+        const menuText =
+            selectedItems.length === 0 && placeholder
+                ? placeholder
+                : `${selectedItems.length} ${selectItemType}`;
```
Comment (kevinbarabash) [inline packages/wonder-blocks-dropdown/components/multi-select-menu.js]: This is tricky since `ngettext` calls should look like:
```
i18n.ngettext("%(num)s Cat", "%(num)s Cats", 3)
```
We should probably be passing in a callback as a prop, something like:
```
<MultiSelectMenu
    ...
    getMenuText={(count) => i18n.ngettext("%(num)s Cat", "%(num)s Cats", count)}
    ...
/>
```

File: packages/wonder-blocks-dropdown/components/multi-select-menu.js [hunk 13]
```diff
@@ -0,0 +1,227 @@
...
+     * Type of the menu items.
+     * For example, if selectItemType is "student" and there are two students
+     * selected, the SelectBox would display "2 students"
+     */
+    selectItemType: string,
+
+    /**
+     * Whether to display shortcuts for Select All and Select None.
+     */
+    displaySelectionShortcuts?: boolean,
```
Comment (kevinbarabash) [inline packages/wonder-blocks-dropdown/components/multi-select-menu.js]: This could be shorter.  Maybe just `shortcuts?: boolean`.

File: packages/wonder-blocks-dropdown/components/multi-select-menu.md (hunk 3 above)
Comment (kevinbarabash) [reply packages/wonder-blocks-dropdown/components/multi-select-menu.md]: This will also simplify some of the other methods in `MultiSelectMenu`.

Comment (wishdasher) [reply packages/wonder-blocks-dropdown/components/multi-select-menu.md]: I'm not sure what you mean here. `changes` is an array of items that have changed states. Do you mean that `changes` should return an array of all the selected items?

File: packages/wonder-blocks-dropdown/utils/types.js (hunk 10 above)
Comment (wishdasher) [reply packages/wonder-blocks-dropdown/utils/types.js]: Some other `type` files in the codebase use `/** ... */`. I will leave it like that to be consistent.

File: packages/wonder-blocks-dropdown/utils/types.js (hunk 11 above)
Comment (wishdasher) [reply packages/wonder-blocks-dropdown/utils/types.js]: This will be used in the upcoming `ActionMenu` pull request. I just didn't want to have to save these somewhere else when I committed the `types.js` file.

File: packages/wonder-blocks-dropdown/components/multi-select-menu.md (hunk 3 above)
Comment (kevinbarabash) [reply packages/wonder-blocks-dropdown/components/multi-select-menu.md]: Yes, if it's just an array of selected items there's a lot of code that can be simplified.  Besides, this is what's being passed as a prop to `MultiSelectMenu`.

File: packages/wonder-blocks-dropdown/utils/types.js (hunk 10 above)
Comment (kevinbarabash) [reply packages/wonder-blocks-dropdown/utils/types.js]: I didn't realize that.

**Comment by wishdasher** [comment]
How the `changes` parameter should work/function is the one unaddressed comment. There is also a bug with `ClickableBehavior` that causes multiple items to be in the hover state. (Look at the menu with selection shortcuts for an example: press `Select all` and then deselect an item.) I will look into this separately.

**Comment by kevinbarabash** [comment]
> press Select all and then deselect an item.) I will look into this separately.

I think it's b/c the "Select all (button)" stops receiving events when it's disable so it doesn't update its state.  That shouldn't be too hard to fix.

File: packages/wonder-blocks-dropdown/components/multi-select-menu.js (hunk 12 above)
Comment (kevinbarabash) [reply packages/wonder-blocks-dropdown/components/multi-select-menu.js]: That's probably a good idea.  It might be best to have `webapp` pass in strings instead of adding `i18n` functions to `wonder-blocks`.

File: packages/wonder-blocks-dropdown/components/multi-select-menu.md (hunk 6 above)
Comment (kevinbarabash) [reply packages/wonder-blocks-dropdown/components/multi-select-menu.md]: The regular menu items look good now, but the shortcuts still have a slight white border on the left/right.

File: packages/wonder-blocks-dropdown/components/multi-select-menu.md (hunk 8 above)
Comment (kevinbarabash) [reply packages/wonder-blocks-dropdown/components/multi-select-menu.md]: Friendly ping. 🙂

**Comment by kevinbarabash** [comment]
Here's what I'm seeing in Safari with the shortcuts.
screen shot 2018-07-03 at 4 59 20 pm
